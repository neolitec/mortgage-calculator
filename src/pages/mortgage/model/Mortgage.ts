import { chunk, sumBy } from 'lodash-es'
import { nanoid } from 'nanoid/non-secure'

export enum PaymentFrequency {
  Monthly = 'monthly',
  Weekly = 'weekly',
  BiWeekly = 'biweekly',
}

export const PaymentsPerYear: Record<PaymentFrequency, number> = {
  [PaymentFrequency.Monthly]: 12,
  // [PaymentFrequency.Weekly]: (365 * 4 + 1) / 4 / 7,
  [PaymentFrequency.Weekly]: 52,
  [PaymentFrequency.BiWeekly]: 26,
}

export interface IMortgage {
  /**
   * Technical id of a mortgage
   */
  id: string
  /**
   * The name of the calculation/simulation
   */
  name: string
  /**
   * The loan amount
   */
  amount: number
  /**
   * Interest rate in percentage
   */
  interestRate: number
  /**
   * Amortization is the number of payments over the life of the loan.
   */
  amortization: number
  /**
   * Payment frequency
   */
  paymentFrequency: PaymentFrequency
}

export interface AmortizationTableRow {
  paymentNumber: number | string
  interest: number
  interestAcc: number
  principal: number
  principalAcc: number
  paid: number
  paidAcc: number
  balance: number
}

class Mortgage implements IMortgage {
  readonly id: string

  private _name = ''

  readonly amount: number

  readonly interestRate: number

  readonly amortization: number

  readonly paymentFrequency: PaymentFrequency = PaymentFrequency.Monthly

  constructor({
    id,
    name,
    amount,
    interestRate,
    amortization,
    paymentFrequency,
  }: Omit<IMortgage, 'id'> & { id?: string }) {
    this.id = id || nanoid()
    this.name = name
    this.amount = amount
    this.interestRate = interestRate
    this.amortization = amortization
    this.paymentFrequency = paymentFrequency
  }

  set name(name: string) {
    // eslint-disable-next-line no-underscore-dangle
    this._name = name
  }

  get name(): string {
    // eslint-disable-next-line no-underscore-dangle
    return this._name
  }

  getPayment(): number {
    const totalPaymentCount = this.getTotalPaymentCount()
    const subdividedInterestRate =
      this.interestRate / 100 / this.getPaymentCountPerYear()
    const top =
      subdividedInterestRate * (1 + subdividedInterestRate) ** totalPaymentCount
    const bottom = (1 + subdividedInterestRate) ** totalPaymentCount - 1
    return this.amount * (top / bottom)
  }

  getTotalPaymentCount(): number {
    return PaymentsPerYear[this.paymentFrequency] * this.amortization
  }

  getPaymentCountPerYear(): number {
    return PaymentsPerYear[this.paymentFrequency]
  }

  getTotalPayment(): number {
    const monthlyPayment = this.getPayment()
    return monthlyPayment * this.getTotalPaymentCount()
  }

  getTotalInterest(): number {
    const totalPayment = this.getTotalPayment()
    const totalInterest = totalPayment - this.amount
    return totalInterest
  }

  getAmortizationTable(
    { perYear } = { perYear: true },
  ): AmortizationTableRow[] {
    const table: AmortizationTableRow[] = []

    const initialRow = {
      paymentNumber: 0,
      interest: 0,
      interestAcc: 0,
      principal: 0,
      principalAcc: 0,
      paid: 0,
      paidAcc: 0,
      balance: this.amount,
    }

    const amortizationTablePerPayment = this.getAmortizationTablePerPayment()

    if (perYear) {
      const perYearAmortizationTable = chunk(
        amortizationTablePerPayment,
        this.getPaymentCountPerYear(),
      ).reduce(
        (acc, yearPayments, index) => {
          const interest = sumBy(yearPayments, 'interest')
          const principal = sumBy(yearPayments, 'principal')

          return [
            ...acc,
            {
              paymentNumber: index + 1,
              interest,
              interestAcc: interest + acc[acc.length - 1].interestAcc,
              principal,
              principalAcc: principal + acc[acc.length - 1].principalAcc,
              paid: interest + principal,
              paidAcc: interest + principal + acc[acc.length - 1].paidAcc,
              balance: yearPayments[yearPayments.length - 1].balance,
            },
          ]
        },
        [initialRow],
      )
      table.push(...perYearAmortizationTable)
    } else {
      table.push(initialRow, ...amortizationTablePerPayment)
    }

    return table
  }

  private getAmortizationTablePerPayment(): AmortizationTableRow[] {
    const table: AmortizationTableRow[] = []

    let prevBalance = this.amount
    let prevInterest = 0
    let prevPrincipal = 0
    for (let i = 0; i < this.getTotalPaymentCount(); i += 1) {
      const interest =
        (prevBalance * this.interestRate) / 100 / this.getPaymentCountPerYear()
      const principal = this.getPayment() - interest
      const balance = prevBalance - principal
      table.push({
        paymentNumber: i + 1,
        interest,
        interestAcc: prevInterest + interest,
        principal,
        principalAcc: prevPrincipal + principal,
        paid: interest + principal,
        paidAcc: prevInterest + interest + prevPrincipal + principal,
        balance: prevBalance - principal,
      })
      prevBalance = balance
      prevInterest = interest
      prevPrincipal = principal
    }

    return table
  }
}

export default Mortgage
