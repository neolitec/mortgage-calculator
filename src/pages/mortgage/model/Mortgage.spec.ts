import type { AmortizationTableRow } from './Mortgage'
import Mortgage, { PaymentFrequency } from './Mortgage'

const A_MORTGAGE = new Mortgage({
  name: 'Mortgage name',
  amount: 288000,
  interestRate: 1.6,
  amortization: 25,
  paymentFrequency: PaymentFrequency.Monthly,
})

const A_WEEKLY_MORTGAGE = new Mortgage({
  name: 'Mortgage name',
  amount: 288000,
  interestRate: 1.6,
  amortization: 25,
  paymentFrequency: PaymentFrequency.Weekly,
})

const A_BIWEEKLY_MORTGAGE = new Mortgage({
  name: 'Mortgage name',
  amount: 288000,
  interestRate: 1.6,
  amortization: 25,
  paymentFrequency: PaymentFrequency.BiWeekly,
})

describe(Mortgage, () => {
  describe('when payment frequency is monthly', () => {
    describe(Mortgage.prototype.getPaymentCountPerYear, () => {
      it('should calculate the number of payments per year', () => {
        const monthlyPayment = A_MORTGAGE.getPaymentCountPerYear()

        expect(monthlyPayment).toEqual(12)
      })
    })

    describe(Mortgage.prototype.getTotalPaymentCount, () => {
      it('should calculate the total number of payments', () => {
        const monthlyPayment = A_MORTGAGE.getTotalPaymentCount()

        expect(monthlyPayment).toEqual(300)
      })
    })

    describe(Mortgage.prototype.getPayment, () => {
      it('should calculate monthly payment', () => {
        const monthlyPayment = A_MORTGAGE.getPayment()

        expect(monthlyPayment).toBeCloseTo(1165.4, 2)
      })
    })

    describe(Mortgage.prototype.getTotalPayment, () => {
      it('should calculate total payment', () => {
        const totalPayment = A_MORTGAGE.getTotalPayment()

        expect(totalPayment).toBeCloseTo(349619.22, 2)
      })
    })

    describe(Mortgage.prototype.getTotalInterest, () => {
      it('should calculate total interest', () => {
        const totalInterest = A_MORTGAGE.getTotalInterest()

        expect(totalInterest).toBeCloseTo(61619.22, 2)
      })
    })

    describe(Mortgage.prototype.getAmortizationTable, () => {
      let amortizationTable: AmortizationTableRow[]

      beforeEach(() => {
        amortizationTable = A_MORTGAGE.getAmortizationTable()
      })

      it('should calculate amortization table', () => {
        expect(amortizationTable).toMatchSnapshot()
      })

      it('should have last element with a null balance', () => {
        const lastElement = amortizationTable[amortizationTable.length - 1]

        expect(lastElement.balance).toBeCloseTo(0, 2)
      })
    })
  })

  describe('when payment frequency is weekly', () => {
    describe(Mortgage.prototype.getPayment, () => {
      it('should calculate monthly payment', () => {
        const monthlyPayment = A_WEEKLY_MORTGAGE.getPayment()

        expect(monthlyPayment).toBeCloseTo(268.09, 2)
      })
    })

    describe(Mortgage.prototype.getTotalPayment, () => {
      it('should calculate total payment', () => {
        const totalPayment = A_WEEKLY_MORTGAGE.getTotalPayment()

        expect(totalPayment).toBeCloseTo(349473.4, 2)
      })
    })

    describe(Mortgage.prototype.getTotalInterest, () => {
      it('should calculate total interest', () => {
        const totalInterest = A_WEEKLY_MORTGAGE.getTotalInterest()

        expect(totalInterest).toBeCloseTo(61473.4, 2)
      })
    })
  })

  describe('when payment frequency is bi-weekly', () => {
    describe(Mortgage.prototype.getPayment, () => {
      it('should calculate monthly payment', () => {
        const monthlyPayment = A_BIWEEKLY_MORTGAGE.getPayment()

        expect(monthlyPayment).toBeCloseTo(536.25, 2)
      })
    })

    describe(Mortgage.prototype.getTotalPayment, () => {
      it('should calculate total payment', () => {
        const totalPayment = A_BIWEEKLY_MORTGAGE.getTotalPayment()

        expect(totalPayment).toBeCloseTo(349516.99, 2)
      })
    })

    describe(Mortgage.prototype.getTotalInterest, () => {
      it('should calculate total interest', () => {
        const totalInterest = A_BIWEEKLY_MORTGAGE.getTotalInterest()

        expect(totalInterest).toBeCloseTo(61516.99, 2)
      })
    })
  })
})
