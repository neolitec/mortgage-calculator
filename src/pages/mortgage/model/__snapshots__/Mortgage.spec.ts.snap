// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mortgage when payment frequency is monthly getAmortizationTable should calculate amortization table 1`] = `
Array [
  Object {
    "balance": 288000,
    "interest": 0,
    "interestAcc": 0,
    "paid": 0,
    "paidAcc": 0,
    "paymentNumber": 0,
    "principal": 0,
    "principalAcc": 0,
  },
  Object {
    "balance": 278554.16174481524,
    "interest": 4538.930497147204,
    "interestAcc": 4538.930497147204,
    "paid": 13984.76875233202,
    "paidAcc": 13984.76875233202,
    "paymentNumber": 1,
    "principal": 9445.838255184815,
    "principalAcc": 9445.838255184815,
  },
  Object {
    "balance": 268956.0768252201,
    "interest": 4386.683832736849,
    "interestAcc": 8925.614329884052,
    "paid": 13984.76875233202,
    "paidAcc": 27969.53750466404,
    "paymentNumber": 2,
    "principal": 9598.08491959517,
    "principalAcc": 19043.923174779986,
  },
  Object {
    "balance": 259203.29135134284,
    "interest": 4231.983278454796,
    "interestAcc": 13157.597608338849,
    "paid": 13984.768752332018,
    "paidAcc": 41954.30625699606,
    "paymentNumber": 3,
    "principal": 9752.785473877222,
    "principalAcc": 28796.708648657208,
  },
  Object {
    "balance": 249293.31188186732,
    "interest": 4074.789282856517,
    "interestAcc": 17232.386891195365,
    "paid": 13984.768752332018,
    "paidAcc": 55939.07500932808,
    "paymentNumber": 4,
    "principal": 9909.979469475502,
    "principalAcc": 38706.68811813271,
  },
  Object {
    "balance": 239223.6047865483,
    "interest": 3915.0616570129764,
    "interestAcc": 21147.44854820834,
    "paid": 13984.76875233202,
    "paidAcc": 69923.8437616601,
    "paymentNumber": 5,
    "principal": 10069.707095319043,
    "principalAcc": 48776.39521345175,
  },
  Object {
    "balance": 228991.595598452,
    "interest": 3752.759564235757,
    "interestAcc": 24900.208112444096,
    "paid": 13984.76875233202,
    "paidAcc": 83908.61251399212,
    "paymentNumber": 6,
    "principal": 10232.009188096263,
    "principalAcc": 59008.40440154802,
  },
  Object {
    "balance": 218594.66835575647,
    "interest": 3587.841509636556,
    "interestAcc": 28488.049622080653,
    "paid": 13984.768752332016,
    "paidAcc": 97893.38126632414,
    "paymentNumber": 7,
    "principal": 10396.927242695461,
    "principalAcc": 69405.33164424347,
  },
  Object {
    "balance": 208030.16493294286,
    "interest": 3420.2653295184286,
    "interestAcc": 31908.31495159908,
    "paid": 13984.768752332018,
    "paidAcc": 111878.15001865616,
    "paymentNumber": 8,
    "principal": 10564.50342281359,
    "principalAcc": 79969.83506705707,
  },
  Object {
    "balance": 197295.38436120684,
    "interest": 3249.988180596013,
    "interestAcc": 35158.30313219509,
    "paid": 13984.76875233202,
    "paidAcc": 125862.91877098818,
    "paymentNumber": 9,
    "principal": 10734.780571736008,
    "principalAcc": 90704.61563879307,
  },
  Object {
    "balance": 186387.58213791685,
    "interest": 3076.966529042026,
    "interestAcc": 38235.269661237115,
    "paid": 13984.76875233202,
    "paidAcc": 139847.6875233202,
    "paymentNumber": 10,
    "principal": 10907.802223289993,
    "principalAcc": 101612.41786208306,
  },
  Object {
    "balance": 175303.96952494205,
    "interest": 2901.1561393572106,
    "interestAcc": 41136.42580059433,
    "paid": 13984.768752332018,
    "paidAcc": 153832.45627565222,
    "paymentNumber": 11,
    "principal": 11083.612612974808,
    "principalAcc": 112696.03047505787,
  },
  Object {
    "balance": 164041.71283567094,
    "interest": 2722.512063060889,
    "interestAcc": 43858.937863655214,
    "paid": 13984.768752332018,
    "paidAcc": 167817.22502798424,
    "paymentNumber": 12,
    "principal": 11262.25668927113,
    "principalAcc": 123958.287164329,
  },
  Object {
    "balance": 152597.93271053815,
    "interest": 2540.9886271992186,
    "interestAcc": 46399.926490854436,
    "paid": 13984.76875233202,
    "paidAcc": 181801.99378031626,
    "paymentNumber": 13,
    "principal": 11443.780125132802,
    "principalAcc": 135402.0672894618,
  },
  Object {
    "balance": 140969.7033808744,
    "interest": 2356.539422668254,
    "interestAcc": 48756.46591352269,
    "paid": 13984.76875233202,
    "paidAcc": 195786.76253264828,
    "paymentNumber": 14,
    "principal": 11628.229329663765,
    "principalAcc": 147030.29661912555,
  },
  Object {
    "balance": 129154.05192089114,
    "interest": 2169.1172923487743,
    "interestAcc": 50925.583205871466,
    "paid": 13984.768752332018,
    "paidAcc": 209771.5312849803,
    "paymentNumber": 15,
    "principal": 11815.651459983244,
    "principalAcc": 158845.94807910878,
  },
  Object {
    "balance": 117147.957487609,
    "interest": 1978.6743190498828,
    "interestAcc": 52904.25752492135,
    "paid": 13984.768752332016,
    "paidAcc": 223756.30003731232,
    "paymentNumber": 16,
    "principal": 12006.094433282135,
    "principalAcc": 170852.0425123909,
  },
  Object {
    "balance": 104948.35054853527,
    "interest": 1785.1618132582848,
    "interestAcc": 54689.419338179636,
    "paid": 13984.768752332018,
    "paidAcc": 237741.06878964434,
    "paymentNumber": 17,
    "principal": 12199.606939073734,
    "principalAcc": 183051.64945146465,
  },
  Object {
    "balance": 92552.11209689334,
    "interest": 1588.5303006901001,
    "interestAcc": 56277.94963886974,
    "paid": 13984.768752332018,
    "paidAcc": 251725.83754197636,
    "paymentNumber": 18,
    "principal": 12396.238451641919,
    "principalAcc": 195447.88790310657,
  },
  Object {
    "balance": 79956.07285420336,
    "interest": 1388.7295096420478,
    "interestAcc": 57666.67914851179,
    "paid": 13984.76875233202,
    "paidAcc": 265710.6062943084,
    "paymentNumber": 19,
    "principal": 12596.039242689973,
    "principalAcc": 208043.92714579654,
  },
  Object {
    "balance": 67157.01246001007,
    "interest": 1185.7083581387524,
    "interestAcc": 58852.38750665054,
    "paid": 13984.76875233202,
    "paidAcc": 279695.3750466404,
    "paymentNumber": 20,
    "principal": 12799.060394193268,
    "principalAcc": 220842.98753998982,
  },
  Object {
    "balance": 54151.658648550954,
    "interest": 979.4149408728968,
    "interestAcc": 59831.80244752344,
    "paid": 13984.768752332022,
    "paidAcc": 293680.1437989724,
    "paymentNumber": 21,
    "principal": 13005.353811459125,
    "principalAcc": 233848.34135144894,
  },
  Object {
    "balance": 40936.68641215382,
    "interest": 769.7965159348739,
    "interestAcc": 60601.59896345832,
    "paid": 13984.768752332018,
    "paidAcc": 307664.91255130444,
    "paymentNumber": 22,
    "principal": 13214.972236397145,
    "principalAcc": 247063.3135878461,
  },
  Object {
    "balance": 27508.71715115036,
    "interest": 556.799491328553,
    "interestAcc": 61158.39845478687,
    "paid": 13984.768752332018,
    "paidAcc": 321649.68130363646,
    "paymentNumber": 23,
    "principal": 13427.969261003465,
    "principalAcc": 260491.28284884954,
  },
  Object {
    "balance": 13864.31781008804,
    "interest": 340.36941126970555,
    "interestAcc": 61498.76786605657,
    "paid": 13984.768752332018,
    "paidAcc": 335634.4500559685,
    "paymentNumber": 24,
    "principal": 13644.399341062313,
    "principalAcc": 274135.68218991184,
  },
  Object {
    "balance": 1.9614617485785857e-8,
    "interest": 120.45094226359461,
    "interestAcc": 61619.21880832017,
    "paid": 13984.768752332018,
    "paidAcc": 349619.2188083005,
    "paymentNumber": 25,
    "principal": 13864.317810068424,
    "principalAcc": 287999.99999998027,
  },
]
`;
